src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs

```diff
@@ -0,0 +1,197 @@
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(DoNotGuardSetAddOrRemoveByContainsDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!TryGetRequiredMethods(context.Compilation, out var containsMethod, out var addMethod, out var removeMethod))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context => OnConditional(context, containsMethod, addMethod, removeMethod), OperationKind.Conditional);
+        }
+
+        private static void OnConditional(OperationAnalysisContext context, IMethodSymbol containsMethod, IMethodSymbol addMethod, IMethodSymbol removeMethod)
+        {
+            var conditional = (IConditionalOperation)context.Operation;
+
+            if (!TryExtractContainsInvocation(conditional.Condition, containsMethod, out var containsInvocation, out var containsNegated))
+            {
+                return;
+            }
+
+            if (!TryExtractAddOrRemoveInvocation(conditional.WhenTrue.Children, addMethod, removeMethod, containsNegated, out var addOrRemoveInvocation))
+            {
+                return;
+            }
+
+            if (!AreInvocationsOnSameInstance(containsInvocation, addOrRemoveInvocation))
+            {
+                return;
+            }
+
+            using var locations = ArrayBuilder<Location>.GetInstance(2);
+            locations.Add(conditional.Syntax.GetLocation());
+            locations.Add(addOrRemoveInvocation.Syntax.Parent!.GetLocation());
+
+            context.ReportDiagnostic(containsInvocation.CreateDiagnostic(Rule, additionalLocations: locations.ToImmutable(), null));
+        }
+
+        private static bool TryGetRequiredMethods(
+            Compilation compilation,
+            [NotNullWhen(true)] out IMethodSymbol? containsMethod,
+            [NotNullWhen(true)] out IMethodSymbol? addMethod,
+            [NotNullWhen(true)] out IMethodSymbol? removeMethod)
+        {
+            var iSetType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericISet1);
+            var iCollectionType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericICollection1);
```

Comment:
`ImmutableHashSet.Builder` and `ImmutableSortedSet.Builder` work (added a test in cf49578).
I also added a test for the expected behavior for `Add` and `Delete` from `IImmutableSet<T>` (only reporting a diagnostic).

But I couldn't get it to work yet, the WIP commit is 4f62beb.

For some reason the following fails when `method` is `System.Collections.Immutable.ImmutableHashSet<string>.Add(string)` and `interfaceMethod` is `System.Collections.Immutable.IImmutableSet<T>.Add(T)`:

```C#
private static bool DoesImplementInterfaceMethod(IMethodSymbol method, IMethodSymbol interfaceMethod)
{
    return method.IsImplementationOfInterfaceMethod(method.Parameters[0].Type, interfaceMethod.ContainingType, interfaceMethod.Name);
}
```

I stepped into `IsImplementationOfInterfaceMethod` which lead to this:
![image](https://github.com/dotnet/roslyn-analyzers/assets/4247811/08a57975-4f15-47b1-90be-a951857ade87)

I don't know why the symbol returned by `FindImplementationForInterfaceMember` is `System.Collections.Immutable.ImmutableHashSet<string>.System.Collections.Immutable.IImmutableSet<string>.Add(string)`. I would except `System.Collections.Immutable.ImmutableHashSet<string>.Add(string)`. This will result in a false comparison result.

Reply:
> ImmutableHashSet.Builder and ImmutableSortedSet.Builder work (added a test in https://github.com/dotnet/roslyn-analyzers/commit/cf4957820e06c8f440c382ac015c590a1b10abfd).
> I also added a test for the expected behavior for Add and Delete from IImmutableSet<T> (only reporting a diagnostic).

Thank you!

> For some reason the following fails when method is `System.Collections.Immutable.ImmutableHashSet<string>.Add(string)` and interfaceMethod is `System.Collections.Immutable.IImmutableSet<T>.Add(T)`:

Because `System.Collections.Immutable.ImmutableHashSet<string>.Add(string)` is not implemented `System.Collections.Immutable.IImmutableSet<T>.Add(T)` instead it has explicit implementation of it instead https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablehashset-1.system-collections-immutable-iimmutableset-t--add?view=net-7.0#system-collections-immutable-immutablehashset-1-system-collections-immutable-iimmutableset((t))-add(-0)

Reply:
I replaced the call to check interface implementations with checking the original definitions (which also contain explicit interface implementations). All tests, including the ones checking for `ImmutableHashSet` and `ImmutableSortedSet`, pass with the following change:

```C#
private static bool DoesImplementInterfaceMethod(IMethodSymbol method, IMethodSymbol interfaceMethod)
{
    var typedInterface = interfaceMethod.ContainingType.Construct(method.Parameters[0].Type);
    var typedInterfaceMethod = typedInterface.GetMembers(interfaceMethod.Name).FirstOrDefault();

    // Check against all original definitions to also cover external interface implementations
    return method.GetOriginalDefinitions().Any(definition => SymbolEqualityComparer.Default.Equals(definition, typedInterfaceMethod));
}
```